### PR TITLE
WorkLock: `restake` lock while claiming tokens

### PR DESCRIPTION
**Pull request by vzotova**
* special method in `StakingEscrow` for depositing tokens from `WorkLock`
* force re-stake lock in `claim` and `batchDeposit`
* removed one slashing test from intercontract test because it's hard to maintain and there is already one test fro this there
* `onlyOwner` modifier for `batchDeposit` method to protect force lock from malicious using

### CODE REVIEW

**Review by cygnusv — APPROVED** [review]

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 1]
```diff
@@ -529,6 +529,52 @@ contract StakingEscrow is Issuer, IERC900History {
...
+    )
+        internal
+    {
+        // reset bit when `reStake` is already disabled
+        if (_info.flags.bitSet(RE_STAKE_DISABLED_INDEX) == true) {
+            _info.flags = _info.flags.toggleBit(RE_STAKE_DISABLED_INDEX);
+            emit ReStakeSet(_staker, true);
+        }
+        // lock `reStake` parameter if it's not locked or locked for too short duration
+        if (_lockReStakeUntilPeriod > _info.lockReStakeUntilPeriod) {
```
Comment (tuxxy) [inline nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:551]: It's not `>=`?

File: nucypher/blockchain/eth/agents.py [hunk 2]
```diff
@@ -489,7 +490,8 @@ def batch_deposit(self,
         if gas_limit and gas_limit < min_gas_batch_deposit:
             raise ValueError(f"{gas_limit} is not enough gas for any batch deposit")
 
-        contract_function: ContractFunction = self.contract.functions.batchDeposit(stakers, number_of_substakes, amounts, lock_periods)
+        contract_function: ContractFunction = self.contract.functions.batchDeposit(
+            stakers, number_of_substakes, amounts, lock_periods, release_period)
```
Comment (tuxxy) [inline nucypher/blockchain/eth/agents.py:494]: :+1:

File: nucypher/blockchain/eth/actors.py [hunk 3]
```diff
@@ -431,7 +432,7 @@ def batch_deposits(self,
         if interactive and not emitter:
             raise ValueError("'emitter' is a required keyword argument when interactive is True.")
 
-        allocator = Allocator(allocation_data_filepath, self.registry, self.deployer_address)
+        allocator = Allocator(allocation_data_filepath, self.registry, self.deployer_address, release_period)
```
Comment (tuxxy) [inline nucypher/blockchain/eth/actors.py:435]: :+1:

File: nucypher/cli/commands/deploy.py [hunk 4]
```diff
@@ -509,14 +509,18 @@ def contracts(general_config, actor_options, mode, activate, gas, ignore_deploye
 @group_general_config
 @group_actor_options
 @click.option('--allocation-infile', help="Input path for token allocation JSON file", type=EXISTING_READABLE_FILE)
+@click.option('--release-period', help="Period number when re-stake lock will be released", type=click.INT)
```
Comment (tuxxy) [inline nucypher/cli/commands/deploy.py:512]: :+1:

File: tests/contracts/integration/test_intercontract_integration.py [hunk 5]
```diff
@@ -329,10 +344,13 @@ def test_batch_deposit(testerchain, token_economics, token, escrow):
     assert escrow.functions.getLockedTokens(staker1, duration).call() == staker1_tokens
     assert escrow.functions.getLockedTokens(staker1, duration + 1).call() == 0
 
+    _wind_down, re_stake, _measure_work, _snapshots = escrow.functions.getFlags(staker1).call()
+    assert re_stake
+    assert escrow.functions.isReStakeLocked(staker1).call()
```
Comment (tuxxy) [inline tests/contracts/integration/test_intercontract_integration.py:349]: :+1:

File: tests/contracts/integration/test_intercontract_integration.py [hunk 6]
```diff
@@ -1192,54 +1201,6 @@ def test_upgrading_adjudicator(testerchain,
     assert adjudicator_v1 == adjudicator.functions.target().call()
 
 
-def test_another_slashing(testerchain,
```
Comment (tuxxy) [inline tests/contracts/integration/test_intercontract_integration.py:1195]: Why was this removed?

File: tests/contracts/main/staking_escrow/test_staking_escrow.py [hunk 7]
```diff
@@ -1532,3 +1614,231 @@ def test_batch_deposit(testerchain, token, escrow_contract, deploy_contract):
     assert event_args['value'] == value3
     assert event_args['firstPeriod'] == current_period + 1
     assert event_args['periods'] == duration3
+
+    event_args = re_stake_lock_events[7]['args']
+    assert event_args['staker'] == staker
+    assert event_args['lockUntilPeriod'] == current_period + 3
+
+
+def test_staking_from_worklock(testerchain, token, escrow_contract, token_economics, deploy_contract):
```
Comment (tuxxy) [inline tests/contracts/main/staking_escrow/test_staking_escrow.py:1623]: :+1:

**Comment by tuxxy** [comment]
Just looked over the `release_period` stuff and how it's added through the contracts. Looks good. I noticed one little detail with some of the `>=` logic.

For me to finish my review, I need to go over the tests in more detail. They're quite extensive. :)

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol (hunk 1 above)
Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:551]: if values are already equal then we can skip this step. For example, user locks restake by themself for long duration, which equal 6 month, then we don't want spend additional gas to change it

File: tests/contracts/integration/test_intercontract_integration.py (hunk 6 above)
Comment (vzotova) [reply tests/contracts/integration/test_intercontract_integration.py:1195]: Really hard to maintain (in this PR hard to make it alive), also we have this test in contracts unit tests.
But if it's an issue - let's open one

Comment (tuxxy) [reply tests/contracts/integration/test_intercontract_integration.py:1195]: Dunno, @cygnusv what are your thoughts?

File: nucypher/cli/commands/deploy.py [hunk 8]
```diff
@@ -509,14 +509,18 @@ def contracts(general_config, actor_options, mode, activate, gas, ignore_deploye
...
 @option_gas
-def allocations(general_config, actor_options, allocation_infile, gas):
+def allocations(general_config, actor_options, allocation_infile, gas, release_period):
     """Deposit stake allocations in batches"""
     emitter = general_config.emitter
     ADMINISTRATOR, _, deployer_interface, local_registry = actor_options.create_actor(emitter)
     if not allocation_infile:
         allocation_infile = click.prompt(PROMPT_FOR_ALLOCATION_DATA_FILEPATH)
+    if release_period is None:
+        release_period = click.prompt(PROMPT_FOR_RELEASE_PERIOD)
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/cli/commands/deploy.py:521]: we can do that, if we really need this - let's open an issue

[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/cli/commands/deploy.py:521]: Filed #2194 but you and @cygnusv are best to judge whether it is needed/nice to have.

**Review by tuxxy — APPROVED** [review]
